File: Sources/JetStreamRpc/Transport.swift
// JetStream RPC — Transport
// Copyright (c) 2024, Sevki <s@sevki.io>
// SPDX-License-Identifier: BSD-3-Clause

// r[impl jetstream.rpc.swift.transport]

import Foundation
import JetStreamWireFormat

/// Transport protocol for sending and receiving frames.
public protocol Transport<TReq, TRes>: Sendable where TReq: Framer, TRes: Framer {
    associatedtype TReq: Framer
    associatedtype TRes: Framer
    func send(_ frame: Frame<TReq>) async throws
    func receive() -> AsyncThrowingStream<Frame<TRes>, Error>
    func close() async throws
}

File: packages/jetstream_rpc/src/transport.ts
/**
 * r[impl jetstream.rpc.ts.transport]
 */
import type { Frame, Framer } from './frame.js';

export interface Transport<TReq extends Framer, TRes extends Framer> {
  send(frame: Frame<TReq>): Promise<void>;
  receive(): AsyncIterable<Frame<TRes>>;
  close(): Promise<void>;
}
